src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -895,6 +895,510 @@ uct_ib_mlx5_devx_dereg_keys(uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mem_t *memh)
+        *umr = NULL;
+        return status;
+    }
+
+    ucs_trace("%s: registered addr %p length %zu lkey 0x%x to "
+              UCT_IB_MLX5_UMR_FMT, uct_ib_device_name(&md->super.dev),
+              ib_mr->addr, ib_mr->length, ib_mr->lkey,
+              UCT_IB_MLX5_UMR_ARG(*umr));
+
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE uint64_t
+uct_ib_mlx5_devx_umr_mkey_hash_code(const uct_ib_md_packed_mkey_t *packed_mkey)
+{
+    return (uint64_t)packed_mkey->vhca_id << 32 | packed_mkey->lkey;
+}
+
+/*
+ * Indicate that given mkey is an UMR mkey and XGVMI enabled.
+ */
+static UCS_F_ALWAYS_INLINE int uct_ib_mlx5_devx_umr_mkey_is_xgvmi(uint32_t lkey)
+{
+    return (UCT_IB_MLX5_MKEY_TAG_UMR == (lkey & 0xFF));
+}
+
+/*
+ * Mark UMR mkey as XGVMI enabled by setting mkey tag (8 LSB in mkey).
+ * This serves for 2 purposes:
+ *  1. On the exporter side - to identify that the mkey is XGVMI enabled, and
+ *     therefore avoiding repetitive XGVMI enablement.
+ *  2. On the importer side - to identify that received rkey is UMR based, and
+ *     therefore should be looked up and stored in UMR alias hash map
+ */
+static UCS_F_ALWAYS_INLINE uint32_t
+uct_ib_mlx5_devx_umr_mkey_set_xgvmi(uct_ib_mlx5_devx_umr_t *umr)
+{
+    umr->umr_mkey->lkey |= UCT_IB_MLX5_MKEY_TAG_UMR;
+    umr->umr_mkey->rkey |= UCT_IB_MLX5_MKEY_TAG_UMR;
+    return umr->umr_mkey->lkey;
+}
+
+static ucs_status_t
+uct_ib_mlx5_devx_umr_mkey_hash_find(uct_ib_mlx5_md_t *md,
+                                    const uct_ib_md_packed_mkey_t *packed_mkey,
+                                    uct_ib_mlx5_devx_umr_alias_t *umr_alias)
+{
+    uint64_t input_key = uct_ib_mlx5_devx_umr_mkey_hash_code(packed_mkey);
+    khint_t k          = kh_get(umr_mkey_map, md->umr_mkey_hash, input_key);
+    if ((k == kh_end(md->umr_mkey_hash)) || (!kh_exist(md->umr_mkey_hash, k))) {
+        return UCS_ERR_NO_ELEM;
+    }
+
+    *umr_alias = kh_val(md->umr_mkey_hash, k);
+    ucs_trace("%s: found " UCT_IB_MLX5_UMR_ALIAS_FMT " for index 0x%x in hash",
+              uct_ib_device_name(&md->super.dev),
+              UCT_IB_MLX5_UMR_ALIAS_ARG(umr_alias),
+              uct_ib_mlx5_mkey_index(packed_mkey->lkey));
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_ib_mlx5_devx_umr_mkey_hash_put(uct_ib_mlx5_md_t *md,
+                                   const uct_ib_md_packed_mkey_t *packed_mkey,
+                                   const uct_ib_mlx5_devx_umr_alias_t *umr_alias)
+{
+    uint64_t input_key = uct_ib_mlx5_devx_umr_mkey_hash_code(packed_mkey);
+    int ret;
+    khint_t k;
+
+    k = kh_put(umr_mkey_map, md->umr_mkey_hash, input_key, &ret);
+    if (ret == UCS_KH_PUT_FAILED) {
+        ucs_error("%s: failed to add " UCT_IB_MLX5_UMR_ALIAS_FMT " to hash map",
+                  uct_ib_device_name(&md->super.dev),
+                  UCT_IB_MLX5_UMR_ALIAS_ARG(umr_alias));
+
+        return UCS_ERR_NO_MEMORY;
+    } else if (ret == UCS_KH_PUT_KEY_PRESENT) {
```

Comment:
no need else

Reply:
Do you mean that the whole `else if (PRESENT){ }` block is not needed?